File: sdk/python/captcha-font-sdk/captcha_font_sdk/font_glyph_slots.py
from __future__ import annotations

from typing import Iterable, List

from .types import FontGlyphFeature, FontGlyphSlot


def align_font_glyph_features_to_slots(
    glyph_features: Iterable[FontGlyphFeature],
    slot_count: int,
    vector_length: int,
) -> List[FontGlyphSlot]:
    if slot_count <= 0:
        raise ValueError("slot_count must be > 0")
    if vector_length <= 0:
        raise ValueError("vector_length must be > 0")

    sorted_features = sorted(glyph_features, key=lambda g: g.rect_index)
    slots: List[FontGlyphSlot] = []
    for slot_index in range(slot_count):
        if slot_index < len(sorted_features):
            glyph = sorted_features[slot_index]
            slots.append(
                FontGlyphSlot(
                    slot_index=slot_index,
                    present=True,
                    rect_index=glyph.rect_index,
                    bbox=glyph.bbox,
                    vector_1d=glyph.vector_1d,
                    density=glyph.density,
                )
            )
        else:
            slots.append(
                FontGlyphSlot(
                    slot_index=slot_index,
                    present=False,
                    rect_index=None,
                    bbox=None,
                    vector_1d=[0] * vector_length,
                    density=0.0,
                )
            )
    return slots

File: typescript-sdk/src/font-glyph-slots.ts
import { FontGlyphFeature, FontGlyphSlot } from './types';

export function alignFontGlyphFeaturesToSlots(
  glyphFeatures: FontGlyphFeature[],
  slotCount: number,
  vectorLength: number
): FontGlyphSlot[] {
  if (slotCount <= 0) {
    throw new Error('slot_count must be > 0');
  }
  if (vectorLength <= 0) {
    throw new Error('vector_length must be > 0');
  }

  const sorted = [...glyphFeatures].sort((a, b) => a.rect_index - b.rect_index);
  const slots: FontGlyphSlot[] = [];
  for (let slotIndex = 0; slotIndex < slotCount; slotIndex += 1) {
    if (slotIndex < sorted.length) {
      const glyph = sorted[slotIndex];
      slots.push({
        slot_index: slotIndex,
        present: true,
        rect_index: glyph.rect_index,
        bbox: glyph.bbox,
        vector_1d: glyph.vector_1d,
        density: glyph.density,
      });
    } else {
      slots.push({
        slot_index: slotIndex,
        present: false,
        rect_index: null,
        bbox: null,
        vector_1d: new Array(vectorLength).fill(0),
        density: 0,
      });
    }
  }
  return slots;
}
